Tests/UnitTests/Core/TrackFitting/KalmanFitterTests.cpp

```diff
@@ -22,560 +20,483 @@
-  mj.visitBackwards(trackTip, [&](const auto& state) {
-    auto typeFlags = state.typeFlags();
-    if (typeFlags.test(TrackStateFlag::OutlierFlag)) {
-      nOutliers++;
-    }
-  });
-  BOOST_CHECK_EQUAL(nOutliers, 1u);
-
-  // Test 7: Run reversed filtering as smoothing
-  kfOptions.reversedFiltering = true;
-  fitRes = kFitter.fit(sourcelinks, rStart, kfOptions);
-  BOOST_CHECK(fitRes.ok());
-  auto fittedWithReversedFiltering = *fitRes;
-  // Check the filtering and smoothing status flag
-  BOOST_CHECK(fittedWithReversedFiltering.reversed);
-  BOOST_CHECK(not fittedWithReversedFiltering.smoothed);
-
-  // Count the number of 'smoothed' states
-  trackTip = fittedWithReversedFiltering.trackTip;
-  mj = fittedWithReversedFiltering.fittedStates;
-  size_t nSmoothed = 0;
-  mj.visitBackwards(trackTip, [&](const auto& state) {
-    if (state.hasSmoothed())
-      nSmoothed++;
-  });
-  BOOST_CHECK_EQUAL(nSmoothed, 6u);
-
-  // Reset to use smoothing formalism
-  kfOptions.reversedFiltering = false;
-
-  // Test 8: Target surface near the tracker exit
-  // Construct a boundless plane surface near the tracker exit
-  Vector3D center(3._m, 0., 0.);
-  Vector3D normal(1., 0., 0.);
-  auto pSurface = Surface::makeShared<PlaneSurface>(center, normal);
-  kfOptions.referenceSurface = pSurface.get();
-  fitRes = kFitter.fit(sourcelinks, rStart, kfOptions);
-  BOOST_CHECK(fitRes.ok());
-
-  // Test 9: Run filtering in outward->inward with starting parameters near the
-  // tracker exit Construct a curvilinear parameters near the tracker exit
-  Vector4D rPos4Outer(3._m, 10_um * gauss(generator), 100_um * gauss(generator),
-                      42_ns);
-  CurvilinearTrackParameters rStartOuter(rPos4Outer, rDir, 1_e / 1_GeV, cov);
-  // Reset the target surface and navigation direction
-  kfOptions.referenceSurface = &rStartOuter.referenceSurface();
-  kfOptions.propagatorPlainOptions.direction = backward;
-  fitRes = kFitter.fit(sourcelinks, rStartOuter, kfOptions);
-  BOOST_CHECK(fitRes.ok());
-
-  // Test 10: Similar to test 9, but run reversed filtering as smoothing
-  kfOptions.reversedFiltering = true;
-  fitRes = kFitter.fit(sourcelinks, rStartOuter, kfOptions);
-  BOOST_CHECK(fitRes.ok());
-
-  // Count the number of 'smoothed' states
-  trackTip = (*fitRes).trackTip;
-  mj = (*fitRes).fittedStates;
-  nSmoothed = 0;
-  mj.visitBackwards(trackTip, [&](const auto& state) {
-    if (state.hasSmoothed())
-      nSmoothed++;
-  });
-  BOOST_CHECK_EQUAL(nSmoothed, 6u);
+  const auto& val = res.value();
+  auto [trackParamsCov, stateRowIndices] =
+      detail::globalTrackParametersCovariance(val.fittedStates, val.trackTip);
+  BOOST_CHECK_EQUAL(trackParamsCov.rows(), sourceLinks.size() * eBoundSize);
+  BOOST_CHECK_EQUAL(stateRowIndices.size(), sourceLinks.size());
+  // TODO where does the reference value come from? can this be computed from
+  // #measurements or something similar?
+  BOOST_CHECK_EQUAL(stateRowIndices.at(val.trackTip), 30);
```

Comment:
To answer the question in the comment, each smoothed track state will have `eBoundSize` rows/cols in the `trackParametersCov`. The `stateRowIndices` is a map of the starting row/indice with the state tip as the key.  Here, the last track state (i.e. the state with the `val.trackTip`) has a starting row/indice = `eBoundSize * (nMeasurements - 1)`, i.e. 6*(6-1)=30.

Reply:
Thanks, I will replace the magic number then.

Reply:
Sorry for the magic number. Thank you for revealing the magic!

Reply:
I fixed it in the test.